### PR TITLE
chore: migrate media-library admin hooks and utils to TS

### PR DESCRIPTION
**Pull request by simotae14**
### What does it do?

- Migrate media-library plugin admin to Typescript
  - fix some Type errors we had in the BE unit tests
  - define the basic TS configuration for the admin part
  - convert all the utils to TS
  - convert all the hooks to TS
  - define the Contracts for the API Requests and Responses

based on the work made in the draft PR https://github.com/strapi/strapi/pull/20867

### Files to migrate?

- [ ] admin
  - [ ] src
    - [ ] components
      - [ ] AssetCard
        - [ ] tests
          - [ ] AssetCardBase.test.jsx
          - [ ] DocAssetCard.test.jsx
          - [ ] ImageAssetCard.test.jsx 
        - [ ] AssetCard.jsx
        - [ ] AssetCardBase.jsx
        - [ ] AudioAssetCard.jsx
        - [ ] AudioPreview.jsx
        - [ ] DocAssetCard.jsx
        - [ ] ImageAssetCard.jsx
        - [ ] UploadingAssetCard.jsx
        - [ ] VideoAssetCard.jsx
        - [ ] VideoPreview.jsx 
      - [ ] AssetDialog
        - [ ] BrowseStep
          - [ ] PaginationFooter
            - [ ] components.jsx
            - [ ] index.jsx
            - [ ] Pagination.jsx
            - [ ] PaginationContext.jsx  
        - [ ] SearchAsset
          - [ ] tests
            - [ ] index.test.jsx
          - [ ] index.jsx
        - [ ] tests
          - [ ] __snapshots__
          - [ ] index.test.jsx
        - [ ] utils
          - [ ] tests
            - [ ] isSelectable.test.js
          - [ ] isSelectable.js     
        - [ ] Filters.jsx
        - [ ] index.jsx
        - [ ] PageSize.jsx 
        - [ ] SelectedStep
          - [ ] index.jsx
        - [ ] tests
          - [ ] AssetDialog.test.jsx
        - [ ] DialogFooter.jsx
        - [ ] index.jsx  
      - [ ] AssetGridList
        - [ ] tests
          - [ ] __snapshots__
          - [ ] AssetGridList.test.jsx 
        - [ ] Draggable.jsx
        - [ ] index.jsx 
      - [ ] Breadcrumbs
        - [ ] tests
          - [ ] __snapshots__
          - [ ] index.test.jsx 
        - [ ] Breadcrumbs.jsx
        - [ ] CrumbSimpleMenuAsync.jsx
        - [ ] index.jsx
      - [ ] BulkMoveDialog
        - [ ] tests
          - [ ] __snapshots__
          - [ ] BulkMoveDialog.test.jsx 
        - [ ] BulkMoveDialog.jsx
        - [ ] index.jsx 
      - [ ] ContextInfo
        - [ ] tests
          - [ ] __snapshots__
          - [ ] ContextInfo.test.jsx 
        - [ ] ContextInfo.jsx
        - [ ] index.jsx 
      - [ ] CopyLinkButton
        - [ ] index.jsx 
      - [ ] EditAssetDialog
        - [ ] PreviewBox
          - [ ] AssetPreview.jsx
          - [ ] components.jsx
          - [ ] CroppingActions.jsx
          - [ ] index.jsx
        - [ ] tests
          - [ ] __snapshots__
          - [ ] EditAssetDialog.test.jsx
          - [ ] index.test.jsx
          - [ ] RemoveAssetDialog.test.jsx
        - [ ] DialogHeader.jsx
        - [ ] index.jsx
        - [ ] RemoveAssetDialog.jsx
        - [ ] ReplaceMediaButton.jsx     
      - [ ] EditFolderDialog
        - [ ] ModalHeader
          - [ ] index.jsx
          - [ ] ModalHeader.jsx
        - [ ] tests
          - [ ]  __snapshots__
          - [ ] EditFolderDialog.test.jsx
        - [ ] EditFolderDialog.jsx
        - [ ] index.jsx
        - [ ] RemoveFolderDialog.jsx
      - [ ] EmptyAssets
        - [ ] EmptyAssetGrid.jsx
        - [ ] index.jsx
      - [ ] FilterList
        - [ ] tests
          - [ ]  __snapshots__
          - [ ] index.test.js
        - [ ] FilterTag.jsx
        - [ ] index.jsx  
      - [ ] FilterPopover
        - [ ] utils
          - [ ] getFilterList.js
        - [ ] FilterValueInput.jsx
        - [ ] index.jsx 
      - [ ] FolderCard
        - [ ] contexts
        - [ ] FolderCard
          - [ ] FolderCard.jsx
          - [ ] index.jsx 
        - [ ] FolderCardBody
          - [ ] FolderCardBody.jsx
          - [ ] index.jsx
        - [ ] FolderCardBodyAction
          - [ ] index.js 
        - [ ] FolderCardCheckbox
          - [ ] FolderCardCheckbox.jsx
          - [ ] index.jsx
        - [ ] hooks
          - [ ] tests
            - [ ] useId.test.js
          - [ ] useId.js
        - [ ] tests
          - [ ] __snapshots__
          - [ ] FolderCard.test.jsx 
        - [ ] index.jsx
      - [ ] FolderGridList
        - [ ] tests
          - [ ] __snapshots__
          - [ ] FolderGridList.test.jsx
        - [ ] FolderGridList.jsx
        - [ ] index.jsx 
      - [ ] MediaLibraryDialog
        - [ ] index.jsx 
      - [ ] MediaLibraryInput
        - [ ] Carousel
          - [ ] tests
            - [ ] CarouselAssets.test.jsx
          - [ ] CarouselAsset.jsx
          - [ ] CarouselAssetActions.jsx
          - [ ] CarouselAssets.jsx
          - [ ] EmptyStateAsset.jsx 
        - [ ] tests
          - [ ] MediaLibraryInput.test.jsx 
        - [ ] index.jsx
      - [ ] SelectTree
        - [ ] tests
          - [ ] __snapshots__
          - [ ] SelectTree.test.jsx
        - [ ] utils
          - [ ] tests
            - [ ] __snapshots__
            - [x] flattenTree.test.js
            - [ ] getOpenValues.test.js
            - [ ] getValuesToClose.test.js
          - [x] flattenTree.js
          - [ ] getOpenValues.js
          - [ ] getValuesToClose.js
        - [ ] index.jsx
        - [ ] Option.jsx
        - [ ] SelectTree.jsx
        - [ ] SelectTree.stories.mdx 
      - [ ] SortPicker
        - [ ] index.jsx
      - [ ] TableList
        - [ ] tests
          - [ ] __snapshots__
          - [ ] CellContent.test.jsx
          - [ ] TableList.test.jsx
          - [ ] TableRows.test.jsx
        - [ ] CellContent.jsx
        - [ ] index.jsx
        - [ ] PreviewCell.jsx
        - [ ] TableRows.jsx
      - [ ] UploadAssetDialog
        - [ ] AddAssetStep
          - [ ] tests
            - [ ] __snapshots__
            - [ ] FromComputerForm.test.jsx
          - [ ] AddAssetStep.jsx
          - [ ] FromComputerForm.jsx
          - [ ] FromUrlForm.jsx 
        - [ ] PendingAssetStep
          - [ ] tests
            - [ ] __snapshots__
            - [ ] PendingAssetStep.test.jsx
          - [ ] PendingAssetStep.jsx 
        - [ ] tests
          - [ ] UploadAssetDialog.test.jsx
        - [ ] UploadAssetDialog.jsx 
      - [ ] UploadProgress
        - [ ] tests
          - [ ] index.test.jsx 
        - [ ] index.jsx
    - [x] hooks
      - [x] __mocks
        - [x] useAssets.js
        - [x] useBulkMove.js
        - [x] useBulkRemove.js
        - [x] useConfig.js
        - [x] useEditFolder.js
        - [x] useFolder.js
        - [x] useFolders.js
        - [x] useFolderStructure.js
        - [x] useMediaLibraryPermissions.js
        - [x] useModalQueryParams.js 
      - [x] tests
        - [x] useAssets.test.js
        - [x] useBulkMove.test.js
        - [x] useBulkRemove.test.js
        - [x] useClipboard.test.ts
        - [x] useConfig.test.js
        - [x] useEditFolder.test.js
        - [x] useFolder.test.js
        - [x] useFolders.test.js
        - [x] useFolderStructure.test.js
        - [x] useModalQueryParams.test.js
        - [x] usePersistentState.test.ts
        - [x] useRemoveAsset.test.js
        - [x] useSelectionState.test.ts 
      - [x] utils
        - [x] tests
          - [x] rename-keys.test.js 
        - [x] rename-keys.js
      - [x] useAssets.js
      - [x] useBulkMove.js
      - [x] useBulkRemove.js
      - [x] useClipboard.ts
      - [x] useConfig.js
      - [x] useCropImg.js
      - [x] useEditAsset.js
      - [x] useEditFolder.js
      - [x] useFolder.js
      - [x] useFolders.js
      - [x] useFolderStructure.js
      - [x] useMediaLibraryPermissions.js
      - [x] useModalQueryParams.js
      - [x] usePersistentState.ts
      - [x] useRemoveAsset.js
      - [x] useSelectionState.ts
      - [x] useUpload.js 
    - [ ] pages
      - [ ] App
        - [ ] components
          - [ ] tests
            - [ ] EmptyOrNoPermissions.test.jsx
          - [ ] BulkActions.jsx
          - [ ] BulkDelete.jsx
          - [ ] BulkMoveButton.jsx
          - [ ] EmptyOrNoPermissions.jsx
          - [ ] Filters.jsx
          - [ ] Header.jsx
        - [ ] ConfigureTheView
          - [ ] components
            - [ ] tests
              - [ ] __snapshots 
            - [ ] Settings.jsx
          - [ ] state
            - [ ] tests
              - [ ] init.test.js
              - [ ] reducer.test.js
            - [ ] actions.js
            - [ ] actionTypes.js
            - [ ] init.js
            - [ ] reducer.js 
          - [ ] tests
            - [ ] __snapshots
            - [ ] ConfigureTheView.test.jsx
          - [ ] index.jsx
        - [ ] MediaLibrary
          - [ ] components
            - [ ] tests
              - [ ] __snapshots__
              - [ ] BulkActions.test.jsx
              - [ ] BulkDeleteButton.test.jsx
              - [ ] BulkMoveButton.test.jsx
              - [ ] EmptyOrNoPermissions.test.jsx
              - [ ] Header.test.jsx
            - [ ] BulkActions.jsx
            - [ ] BulkDeleteButton.jsx
            - [ ] BulkMoveButton.jsx
            - [ ] EmptyOrNoPermissions.jsx
            - [ ] Filters.jsx
            - [ ] Header.jsx
          - [ ] tests
            - [ ] MediaLibrary.test.jsx
          - [ ] index.jsx
        - [ ] index.jsx 
      - [ ] SettingsPage
        - [ ] tests
          - [ ] index.test.jsx
          - [ ] reducer.test.js 
        - [ ] index.jsx
        - [ ] init.js
        - [ ] reducer.js
    - [x] translations
      - [x] tests
        - [x] plural.test.js
    - [x] utils
      - [x] tests
        - [x] appendSearchParamsToUrl.test.js
        - [x] containsAssetFilter.test.js
        - [x] downloadFile.test.js
        - [x] findRecursiveFolderByValue.test.js
        - [x] findRecursiveFolderMetadatas.test.js
        - [x] formatBytes.test.js
        - [x] formatDuration.test.js
        - [x] getAllowedFiles.test.js
        - [x] getAPIInnerErrors.test.js
        - [x] getBreadcrumbDataCM.test.js
        - [x] getBreadcrumbDataML.test.js
        - [x] getFileExtension.test.js
        - [x] getFolderParents.test.js
        - [x] getFolderURL.test.js
        - [x] normalizeAPIError.test.js
        - [x] prefixFileUrlWithBackendUrl.test.js
        - [x] prefixPluginTranslations.test.js
        - [x] toSingularTypes.test.js
        - [x] typeFromMime.test.js
      - [x] appendSearchParamsToUrl.js
      - [x] containsAssetFilter.js
      - [x] createAssetUrl.js
      - [x] displayedFilters.js
      - [x] downloadFilte.js
      - [x] findRecursiveFolderByValue.js
      - [x] findRecursiveFolderMetadatas.js
      - [x] formatBytes.js
      - [x] formatDuration.js
      - [x] getAllowedFiles.js
      - [x] getAPIInnerErrors.js
      - [x] getBreadcrumbDataCM.js
      - [x] getBreadcrumbDataML.js
      - [x] getFileExtension.js
      - [x] getFolderParents.js
      - [x] getFolderURL.js
      - [x] getTrad.js
      - [x] index.js
      - [x] moveElement.js
      - [x] normalizeAPIError.js
      - [x] prefixFileUrlWithBackendUrl.js
      - [x] prefixPluginTranslations.js
      - [x] rawFileToAsset.js
      - [x] toSingularTypes.js
      - [x] typeFromMime.js
      - [x] urlsToAssets.js
      - [x] urlYupSchema.js
    - [ ] constants.js
    - [ ] index.js
    - [x] pluginId.js 
  - [x] tests  
    - [x] handlers.js
    - [x] setup.js
    - [x] utils.jsx 
- [ ] documentation 
  - [ ] content-api.json
- [x] .eslintignore
- [x] .eslintrc
- [x] .gitattributes
- [x] .gitignore
- [x] .npmignore
- [x] jest.config.front.js
- [x] jest.config.js
- [x] jsconfig.json
- [x] LICENSE
- [ ] package.json
- [ ] packup.config.ts
- [x] README.md
- [x] strapi-server.js 

Define the 
 - [ ] Contracts

### CODE REVIEW

**Review by Marc-Roig — DISMISSED** [review]

File: packages/core/upload/admin/src/utils/tests/appendSearchParamsToUrl.test.ts [hunk 1]
```diff
@@ -14,92 +14,88 @@ describe('appendSearchParamsToUrl', () => {
     const url = 'https://example.com/';
 
     test('returns original string with no update time', () => {
-      expect(appendSearchParamsToUrl({ url, params: undefined })).toMatchInlineSnapshot(
-        `"https://example.com/"`
+      expect(appendSearchParamsToUrl({ url, params: undefined })).toBe('https://example.com/');
+      // @ts-ignore
```
Comment (remidej) [inline packages/core/upload/admin/src/utils/tests/appendSearchParamsToUrl.test.ts:18]: can you add a reason why it's needed in the comment when you do ts-ignore?

File: packages/core/upload/admin/src/utils/tests/downloadFile.test.ts [hunk 2]
```diff
@@ -10,7 +10,8 @@ describe('downloadFile', () => {
 
     documentSpy.mockReturnValue({
       click: clickSpy,
-      set href(val) {
+      // @ts-ignore I don't need to change the type of the mocked element
+      set href(val: any) {
```
Comment (remidej) [inline packages/core/upload/admin/src/utils/tests/downloadFile.test.ts:14]: what's the error here? I feel like we could avoid ignoring TS by using unknown and adding a typeof check

File: packages/core/upload/admin/src/utils/tests/normalizeAPIError.test.ts [hunk 3]
```diff
@@ -35,7 +34,6 @@ const API_APPLICATION_ERROR_FIXTURE = new FetchError('ApplicationError', {
       details: {},
     },
   },
-  status: 400,
```
Comment (remidej) [inline packages/core/upload/admin/src/utils/tests/normalizeAPIError.test.ts:38]: Why are we removing the status btw?

File: packages/core/upload/admin/tests/handlers.ts [hunk 4]
```diff
@@ -183,13 +190,13 @@ const handlers = [
   }),
 
   rest.get('*/an-image.png', (req, res, ctx) =>
-    res(ctx.set('Content-Type', 'image/png'), ctx.body())
+    res(ctx.set('Content-Type', 'image/png'), ctx.body('Successfull response'))
```
Comment (remidej) [inline packages/core/upload/admin/tests/handlers.ts:193]: successful just has one L

File: packages/core/upload/shared/contracts/folders.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  locale: string | null;
+  children?: {
+    count: number;
+  };
+  files?: {
+    count: number;
+  };
+}
+
+export interface FolderEnriched extends Folder {
```
Comment (remidej) [inline packages/core/upload/shared/contracts/folders.ts:55]: Could you find a name that indicates in which way it's enriched? Or where it's intended to be used? Enriched could mean lots of things

**Pull request closed without merging**